File: pe/visualstudio/PE.CSharp/Problem0003.cs
using System.Collections.Generic;
using System.Linq;

namespace PE.CSharp
{
    public class Problem0003
    {
        private int _factorBy;
        private int _currentNumberToFactor;
        private HashSet<int> _listOfPrimeFactors; 
        
        public int GreatestPrimeFactor(int numberToFactor)
        {
            return PrimeFactors(numberToFactor).Max();
        }

        public IEnumerable<int> PrimeFactors(int numberToFactor)
        {
            _factorBy = 2;
            _currentNumberToFactor = numberToFactor;
            _listOfPrimeFactors = new HashSet<int>();

            while (IsFactorByLessThanTheNumberToFactor())
            {
                DetermineIfAFactorAndIncrement();
            }

            return _listOfPrimeFactors;
        }

        private bool IsFactorByLessThanTheNumberToFactor()
        {
            return _factorBy <= _currentNumberToFactor;
        }

        private void DetermineIfAFactorAndIncrement()
        {
            if (IsAFactorOf())
            {
                _listOfPrimeFactors.Add(_factorBy);
                _currentNumberToFactor = _currentNumberToFactor/_factorBy;
            }
            else
            {
                _factorBy++;
            }
        }

        private bool IsAFactorOf()
        {
            return _currentNumberToFactor % _factorBy == 0;
        }
    }
}

File: pe/visualstudio/PE.JavaScript/Implementation/Problem0003.js
var javascriptPE;
(function (javascriptPE) {
    (function (implementation) {

        var problem0003 = (function() {

            function problem0003() {
                
            }

            problem0003.prototype.GreatestPrimeFactor = function(numberToFactor) {
                var factorBy = 2;
                var currentNumberToFactor = numberToFactor;
                var maxPrimeNumber;

                while (factorBy <= currentNumberToFactor) {
                    if (currentNumberToFactor % factorBy == 0) {
                        maxPrimeNumber = factorBy;
                        currentNumberToFactor = currentNumberToFactor / factorBy;
                    } else {
                        factorBy++;
                    }
                }

                return maxPrimeNumber;
            };


            return problem0003;
        })();
        implementation.problem0003 = problem0003;

    })(javascriptPE.implementation || (javascriptPE.implementation = {}));
})(javascriptPE || (javascriptPE = {}));